python/vespa/vespa/package.py

```diff
@@ -468,3 +490,289 @@ def deploy(self, disk_folder: str, container_memory: str = "4G"):
+            "X-Key-Id": self.tenant + ":" + self.application + ":" + "default",
+            "X-Key": self.api_public_key_bytes,
+            "X-Authorization": standard_b64encode(signature),
+            **headers,
+        }
+
+        body.seek(0)
+        self.connection.request(method, path, body, headers)
+        with self.connection.getresponse() as response:
+            parsed = json.load(response)
+            if response.status != 200:
+                raise RuntimeError(
+                    "Status code "
+                    + str(response.status)
+                    + " doing "
+                    + method
+                    + " at "
+                    + url
+                    + ":\n"
+                    + parsed["message"]
+                )
+            return parsed
+
+    def get_dev_region(self) -> str:
+        return self.request("GET", "/zone/v1/environment/dev/default")["name"]
+
+    def get_endpoint(
+        self, instance: str, region: str, application_package_name: str
+    ) -> str:
+        endpoints = self.request(
+            "GET",
+            "/application/v4/tenant/{}/application/{}/instance/{}/environment/dev/region/{}".format(
+                self.tenant, self.application, instance, region
+            ),
+        )["endpoints"]
+        container_url = [
+            endpoint["url"]
+            for endpoint in endpoints
+            if endpoint["cluster"] == "{}_container".format(application_package_name)
+        ]
+        if not container_url:
+            raise RuntimeError("No endpoints found for container 'test_app_container'")
+        return container_url[0]
+
+    def to_application_zip(self) -> BytesIO:
+        buffer = BytesIO()
+        with zipfile.ZipFile(buffer, "a") as zip_archive:
+            zip_archive.writestr(
+                "application/schemas/{}.sd".format(
+                    self.application_package.schema.name
+                ),
+                self.application_package.schema_to_text,
+            )
+            zip_archive.writestr(
+                "application/services.xml", self.application_package.services_to_text
+            )
+            zip_archive.writestr(
+                "application/security/clients.pem",
+                self.data_certificate.public_bytes(serialization.Encoding.PEM),
+            )
+
+        return buffer
+
+    def start_deployment(self, instance: str, job: str, disk_folder: str) -> int:
+        deploy_path = (
+            "/application/v4/tenant/{}/application/{}/instance/{}/deploy/{}".format(
+                self.tenant, self.application, instance, job
+            )
+        )
+
+        application_zip_bytes = self.to_application_zip()
+
+        self.write_private_key_and_cert(
+            self.data_key, self.data_certificate, disk_folder
+        )
+        with open(os.path.join(disk_folder, "application.zip"), "wb") as zipfile:
+            zipfile.write(application_zip_bytes.getvalue())
```

Comment:
This is a good question. It is not necessary but I thought it would be nice to have a copy in case the user wants to check the config files that got deployed.

A future change would include the option to deploy a user generated .zip file to Cloud from pyvespa, instead of building the .zip file from our limited application package API. 

That way assuming there will always be a .zip file in the `disk_folder` would tie those two use cases together. But for now it is mostly for the user convenience of checking the config files.

Reply:
I think that's a mixing of concerns. Also, wouldn't it be better for the user to just inspect the application package object and get the files printed to stdout?

Reply:
I removed the step that write zip file to disk. I can revisit that if needed.